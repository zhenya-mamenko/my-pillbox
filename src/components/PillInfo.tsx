import React, { useEffect, type ComponentProps } from 'react';
import { StyleSheet, Text, View, } from 'react-native';
import { Image } from 'expo-image';
import { GestureHandlerRootView, Gesture, GestureDetector, } from 'react-native-gesture-handler';
import ReanimatedSwipeable from 'react-native-gesture-handler/ReanimatedSwipeable';
import Reanimated, { configureReanimatedLogger, runOnJS, SharedValue, useAnimatedStyle, useSharedValue, } from 'react-native-reanimated';
import * as Haptics from 'expo-haptics';

import t from '@/helpers/localization';
import { Pill } from '@/helpers/types';
import ServiceButton from './ServiceButton';
import useThemeColor from '@/hooks/useThemeColor';


configureReanimatedLogger({ strict: false});

interface PillInfoProps extends ComponentProps<any> {
  imageSize: number;
  pill: Pill;
  onDelete: (pill: Pill) => void;
  onEdit: (pill: Pill) => void;
  onImageSizeChanged: (size: number) => void;
}

const PillInfo: React.FC<PillInfoProps> = ({ imageSize, pill, onDelete, onEdit, onImageSizeChanged, }) => {

  const styles = StyleSheet.create({
    container: {
      alignItems: 'flex-start',
      backgroundColor: useThemeColor('background'),
      borderColor: useThemeColor('borderDim'),
      borderRadius: 10,
      borderWidth: 1,
      marginLeft: 10,
      marginRight: 0,
      marginTop: 10,
      padding: 10,
    },
    image: {
      borderColor: useThemeColor('border'),
      borderWidth: 0.5,
      height: '100%',
      marginBottom: 0,
      marginRight: 20,
      width: '100%',
      zIndex: 100,
    },
    textContainer: {
      flex: 1,
      flexDirection: 'column',
      height: '100%',
      justifyContent: 'space-between',
      width: '100%',
    },
    linesContainer: {
      flexDirection: 'column',
      verticalAlign: 'top',
    },
    title: {
      color: useThemeColor('text'),
      fontSize: 18,
      fontWeight: 'bold',
      marginTop: -5,
    },
    description: {
      color: useThemeColor('text'),
      fontSize: 14,
    },
    bestBefore: {
      alignSelf: 'flex-end',
      color: useThemeColor('textDim'),
      fontSize: 12,
      textAlign: 'right',
    },
    rightAction: {
      alignItems: 'center',
      flex: 1,
      height: 120,
      justifyContent: 'center',
      marginLeft: 0,
      marginTop: 10,
      padding: 10,
      width: 70,
    },
  });

  const { imageData, name, description, bestBefore } = pill;

  function RightAction(prog: SharedValue<number>, drag: SharedValue<number>, swipeableMethods: any) {
    const styleAnimation = useAnimatedStyle(() => {
      return {
        transform: [{ translateX: drag.value + 70 }],
      };
    });

    return (
      <Reanimated.View style={styleAnimation}>
        <View style={styles.rightAction}>
          <ServiceButton icon="edit" onPress={() => { swipeableMethods.close(); onEdit(pill); }} />
          <ServiceButton icon="delete" onPress={() => { swipeableMethods.close(); onDelete(pill); }} />
        </View>
      </Reanimated.View>
    );
  }

  const scale = useSharedValue(1);
  const savedSize = useSharedValue(imageSize);
  useEffect(() => {
    savedSize.value = imageSize;
  }, [imageSize, savedSize]);

  const pinch = Gesture.Pinch()
    .onStart(() => {
      scale.value = 1;
    })
    .onUpdate((e) => {
      scale.value = e.scale * savedSize.value >= 100 && e.scale * savedSize.value <= 320 ? e.scale : scale.value;
    })
    .onEnd(() => {
      savedSize.value = scale.value * savedSize.value;
      scale.value = 1;
      runOnJS(onImageSizeChanged)(savedSize.value);
    });

  const scaleAnimation = useAnimatedStyle(() => {
    return {
      ...styles.image,
      width: scale.value * savedSize.value,
      height: scale.value * savedSize.value,
    };
  });

  const doubleTap = Gesture.Tap()
    .maxDuration(250)
    .numberOfTaps(2)
    .onEnd(() => {
      savedSize.value = 130;
      runOnJS(onImageSizeChanged)(savedSize.value);
    });

  const composed = Gesture.Simultaneous(pinch, doubleTap);

  return (
    <GestureHandlerRootView>
      <ReanimatedSwipeable
        friction={2}
        rightThreshold={40}
        renderRightActions={RightAction}
        onSwipeableOpen={async () => await Haptics.notificationAsync(Haptics.NotificationFeedbackType.Success)}
      >
        <View
          style={
            {...styles.container, ...(imageSize < 210 ? {flexDirection: 'row'} : {flexDirection: 'column'})}
          }
        >
          <GestureDetector gesture={composed}>
            <Reanimated.View style={scaleAnimation}>
              <Image
                source={{ uri: `data:image/jpg;base64,${imageData}` }}
                style={ styles.image }
              />
            </Reanimated.View>
          </GestureDetector>
          <View style={
            {...styles.textContainer, ...(imageSize < 210 ? {marginTop: 0,} : {marginTop: 20,}) }
          }>
            <View style={styles.linesContainer}>
              <Text style={styles.title}>{name}</Text>
              <Text
                ellipsizeMode='tail'
                numberOfLines={Math.round(imageSize / 25)}
                style={styles.description}
              >
                {
                  description
                }
              </Text>
            </View>
            <View>
              <Text style={styles.bestBefore}>
                {bestBefore &&
                  `${t('pill_best_before')}: ${(new Date(bestBefore)).toLocaleDateString()}`
                }
              </Text>
            </View>
          </View>
        </View>
      </ReanimatedSwipeable>
    </GestureHandlerRootView>
  );
};

export default PillInfo;
