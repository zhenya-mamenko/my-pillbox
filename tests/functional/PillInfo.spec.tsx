import '@testing-library/react-native/extend-expect';
import { render, screen, } from '@testing-library/react-native';
import PillInfo from '@/components/PillInfo';


jest.mock('@/hooks/useThemeColor', () => (color: string) => {
  const colors: any = {
    background: '#ffffff',
    border: '#000000',
    borderDim: '#ccccccc',
    secondary: '#555555',
    text: '#000000',
    textDim: '#888888',
  }
  return colors[color];
});

jest.mock('@/helpers/localization', () => ({
  __esModule: true,
  default: (s: string) => s
}));

jest.mock('@/components/ServiceButton', () => {
  return {
    __esModule: true,
    default: ({icon}: {icon: string}) => <div>{icon}</div>,
  };
});

jest.mock('react-native-gesture-handler', () => {
  const GS = jest.requireActual('react-native-gesture-handler');
  return {
    __esModule: true,
    ...GS,
    GestureDetector: ({children}: {children: any}) => <div>{children}</div>,
  };
});

jest.mock('react-native-gesture-handler/ReanimatedSwipeable', () => {
  return {
    __esModule: true,
    default: ({children}: {children: any}) => <div>{children}</div>,
  };
});

jest.mock('react-native-reanimated', () => {
  const RNR = jest.requireActual('react-native-reanimated');
  return {
    __esModule: true,
    ...RNR,
    default: {
      ...RNR.default,
      View: ({children}: {children: any}) => <div>{children}</div>,
    }
  }
});

describe('PillInfo appearence', () => {

  const pill = {
    id: 1,
    name: 'Pill 1',
    description: 'Description',
    bestBefore: '2025-10-01',
    imageData: 'iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII',
  }

  beforeAll(() => {
    jest.spyOn(Date.prototype, 'toLocaleDateString').mockImplementation(() => '2025-10-01');
  });

  it('Should render', () => {
    render(
      <PillInfo
        imageSize={100}
        pill={pill}
        onDelete={() => {}}
        onEdit={() => {}}
        onImageSizeChanged={() => {}}
        onStartGesture={() => {}}
      />
    );
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('Should contain right data', async () => {
    render(
      <PillInfo
        imageSize={100}
        pill={pill}
        onDelete={() => {}}
        onEdit={() => {}}
        onImageSizeChanged={() => {}}
        onStartGesture={() => {}}
      />
    );
    expect(screen.getByTestId('pill-name-1')).toHaveTextContent(pill.name);
    expect(screen.getByTestId('pill-description-1')).toHaveTextContent(pill.description);
    expect(screen.getByTestId('pill-best-before-1')).toHaveTextContent(`pill_best_before: ${(new Date(pill.bestBefore)).toLocaleDateString()}`);
    expect(screen.getByTestId('pill-image-1')).toHaveProp('source', [{ uri: `data:image/jpg;base64,${pill.imageData}` }]);
  });

});
