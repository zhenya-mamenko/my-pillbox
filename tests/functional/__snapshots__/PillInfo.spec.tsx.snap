// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PillInfo appearence Should render 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <div>
    <View
      style={
        {
          "alignItems": "flex-start",
          "backgroundColor": "#ffffff",
          "borderColor": "#ccccccc",
          "borderRadius": 10,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginLeft": 10,
          "marginRight": 0,
          "marginTop": 10,
          "padding": 10,
        }
      }
    >
      <div>
        <div>
          <ViewManagerAdapter_ExpoImage
            borderColor={4278190080}
            borderWidth={0.5}
            containerViewRef={"[React.ref]"}
            contentFit="cover"
            contentPosition={
              {
                "left": "50%",
                "top": "50%",
              }
            }
            height="100%"
            marginBottom={0}
            marginRight={20}
            nativeViewRef={"[React.ref]"}
            onError={[Function]}
            onLoad={[Function]}
            onLoadStart={[Function]}
            onProgress={[Function]}
            placeholder={[]}
            source={
              [
                {
                  "uri": "data:image/jpg;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIAQMAAAD+wSzIAAAABlBMVEX///+/v7+jQ3Y5AAAADklEQVQI12P4AIX8EAgALgAD/aNpbtEAAAAASUVORK5CYII",
                },
              ]
            }
            style={
              {
                "borderColor": "#000000",
                "borderWidth": 0.5,
                "height": "100%",
                "marginBottom": 0,
                "marginRight": 20,
                "width": "100%",
                "zIndex": 100,
              }
            }
            testID="pill-image-1"
            transition={null}
            width="100%"
            zIndex={100}
          />
        </div>
      </div>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
            "marginTop": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            {
              "flexDirection": "column",
              "verticalAlign": "top",
            }
          }
        >
          <Text
            style={
              {
                "color": "#000000",
                "fontSize": 18,
                "fontWeight": "bold",
                "marginTop": -5,
              }
            }
            testID="pill-name-1"
          >
            Pill 1
          </Text>
          <Text
            ellipsizeMode="tail"
            numberOfLines={4}
            style={
              {
                "color": "#000000",
                "fontSize": 14,
              }
            }
            testID="pill-description-1"
          >
            Description
          </Text>
        </View>
        <View>
          <Text
            style={
              {
                "alignSelf": "flex-end",
                "color": "#888888",
                "fontSize": 12,
                "textAlign": "right",
              }
            }
            testID="pill-best-before-1"
          >
            pill_best_before: 2025-10-01
          </Text>
        </View>
      </View>
    </View>
  </div>
</View>
`;
